File: lib/opendmm/engines/av_entertainments.rb
require 'cgi'
require 'opendmm/movie'
require 'opendmm/utils/httparty'

module OpenDMM
  module Engine
    module AvEntertainments
      def self.search(query)
        queries = normalize(query)
        LOGGER.debug queries
        queries.lazy.map do |query|
          begin
            Movie.new(query).details
          rescue StandardError => e
            LOGGER.debug query
            LOGGER.debug e
            nil
          end
        end.find(&:present?)
      end

      private

      def self.normalize(query)
        query.scan(/((?:S2|2D|3D|[a-z]){2,6}?)[-_\s]?(S?\d{2,5})/i).map do |pair|
          alpha = pair[0].upcase
          digit = pair[1]
          case alpha
          when /SMBD/, /CWPBD/
            digit = digit.to_i.to_s.rjust(2, '0')
          end
          "#{alpha}-#{digit}"
        end
      end

      module Site
        include HTTParty
        base_uri 'www.aventertainments.com'
        headers({
          "User-Agent" => "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/35.0.1916.153 Safari/537.36"
        })

        def self.search(query)
          get "/search_Products.aspx?keyword=#{CGI::escape(query)}"
        end
      end

      class Search
        def initialize(query, response)
          @query = query
          @response = response
          @html = Nokogiri.HTML @response
        end

        def result
          @html.at_css('div.main-unit2 > table a')['href']
        end
      end

      class Movie < OpenDMM::Movie
        def initialize(query)
          search = Search.new(query, Site.search(query))
          super(query, Site.get(search.result))

          @details.title       = @html.css('#mini-tabet > h2').text
          @details.cover_image = @html.at_css('#titlebox > div.list-cover > img')['src'].gsub('jacket_images', 'bigcover')
          @details.code        = @html.css('#mini-tabet > div').text.remove('商品番号:')
          @details.categories  = @html.xpath('//*[@id="TabbedPanels1"]/div/div[2]/div[2]//ol').map(&:text)

          @html.css('#titlebox > ul > li').each do |li|
            case li.css('span').text
            when /主演女優/
              @details.actresses = li.css('a').map(&:text)
            when /スタジオ/
              @details.maker = li.css('a').text
            when /シリーズ/
              @details.series = li.css('a').text
            when /発売日/
              @details.release_date = li.text
            when /収録時間/
              @details.movie_length = li.text
            end
          end
        end
      end
    end
  end
end

File: aventertainments.go
package opendmm

import (
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"

	"github.com/junzh0u/httpx"

	"github.com/PuerkitoBio/goquery"
)

func aveEngine(keyword string, wg *sync.WaitGroup, metach chan MovieMeta) {
	urlstr := fmt.Sprintf(
		"http://www.aventertainments.com/search_Products.aspx?keyword=%s",
		url.QueryEscape(keyword),
	)
	doc, err := newDocument(urlstr, httpx.ReadBodyInUTF8(http.Get))
	if err != nil {
		return
	}

	doc.Find("div.main-unit2 > table a").Each(
		func(i int, a *goquery.Selection) {
			href, ok := a.Attr("href")
			if ok {
				wg.Add(1)
				go func() {
					defer wg.Done()
					aveParse(href, keyword, metach)
				}()
			}
		})
}

func aveParse(urlstr string, keyword string, metach chan MovieMeta) {
	doc, err := newDocument(urlstr, httpx.ReadBodyInUTF8(http.Get))
	if err != nil {
		return
	}

	var meta MovieMeta
	var ok bool
	meta.Page = urlstr
	meta.Title = doc.Find("#mini-tabet > h2").Text()
	meta.CoverImage, ok = doc.Find("#titlebox > div.list-cover > img").Attr("src")
	if ok {
		meta.CoverImage = strings.Replace(meta.CoverImage, "jacket_images", "bigcover", -1)
	}
	meta.Code = strings.TrimSpace(strings.Replace(doc.Find("#mini-tabet > div").Text(), "商品番号:", "", -1))
	doc.Find("#titlebox > ul > li").Each(
		func(i int, li *goquery.Selection) {
			k := li.Find("span").Text()
			if strings.Contains(k, "主演女優") {
				meta.Actresses = li.Find("a").Map(
					func(i int, a *goquery.Selection) string {
						return a.Text()
					})
			} else if strings.Contains(k, "スタジオ") {
				meta.Maker = li.Find("a").Text()
			} else if strings.Contains(k, "シリーズ") {
				meta.Series = li.Find("a").Text()
			} else if strings.Contains(k, "発売日") {
				meta.ReleaseDate = li.Text()
			} else if strings.Contains(k, "収録時間") {
				meta.MovieLength = li.Text()
			}
		})

	if codeEquals(keyword, meta.Code) {
		metach <- meta
	}
}
